File: JavaScript/07. Arrays/Arrays-Homework/Arrays-Homework/07. BinarySearch/BinarySearch.js
/*
 * Problem 7. Binary search

Write a script that finds the index of given element in a sorted array of integers by using the binary search algorithm.
 */
var arr = [],
    key,
    finalIndex,
    returnedValue,
    found,
    index;

arr = [11, 1, 3, 4, 5, 66, 9, 0];
console.log("arr: " + arr.join(", "));
key = 11;
console.log("Search for: " + key);
index = parseInt(arr.length / 2);
finalIndex = index;
arr = arr.sort(function (a, b) { return a - b });
console.log("arr(sorted): " + arr.join(", "));
returnedValue = -1;
found = true;

function BinarySearch(arr, key, index) {
    if (key == arr[index]) {
        if (found) {
            returnedValue = finalIndex;
            found = false;
        }
        return;
    }
    if (key < arr[index]) {
        arr = arr.slice(0, index);
        index = parseInt(arr.length / 2);
        finalIndex -= (arr.length - index);
        BinarySearch(arr, key, index);
    }
    if (key > arr[index]) {
        arr = arr.slice(index + 1, arr.length);
        index = parseInt(arr.length / 2);
        finalIndex += (arr.length - index);
        BinarySearch(arr, key, index);
    }
}

BinarySearch(arr, key, index);
console.log("Returned index: " + returnedValue);

File: C# II/02. Multidimensional Arrays/MultidimensionalArrays-Homework/04. BinarySearch/BinarySearch.cs
using System;
using System.Linq;

class BinarySearch
{
    /*
      4. Binary search

        Write a program, that reads from the console an array of N 
        integers and an integer K, sorts the array and using the 
        method Array.BinSearch() finds the largest number in the 
        array which is ≤ K.

    */
    static void Main()
    {
        Console.WriteLine("Enter integer number in the array separated by space: ");
        int[] arr = Console.ReadLine().Split().Select<string, int>(n => int.Parse(n)).ToArray<int>();
        Console.Write("Enter number K: ");
        int k = int.Parse(Console.ReadLine());
        Array.Sort(arr);
        int index = 0;
        int p = 0;
        Console.WriteLine("\ninput: ");
        Console.WriteLine("[ " + string.Join(" | ",arr) + " ]");
        Console.WriteLine("\nresult: ");
        if(k >= arr[0])
        {
            while (true)
            {
                index = Array.BinarySearch(arr, k - p);
                p++;
                if (index >= 0) break;
            }
            Console.WriteLine("number {0} on index {1}", arr[index], index);
        }
        else
        {
            Console.WriteLine("There is no such number! K is < {0} = arr[{1}]", arr[index], index);
        }
        Console.WriteLine();
    }
}

